feat(flex-stacker): Enable the Time-Of-Flight sensors and add GCodes to interface with them.

vegano1 opened the pull request:
# Overview

The TMF8820/21/28 is a dToF (direct time of flight) wide field of view optical distance sensor module achieving up to 5000 mm target detection distance and has 3x3, 4x4, 3x6, or 8x8 zones. It will allow the Flex Stacker to detect if there is labware on the X and Z axis. This pull request brings up both TOF sensors and exposes some functionality via new Gcodes. This pull request goes in conjunction with this monorepo pull request [Opentrons/opentrons#17478](https://github.com/Opentrons/opentrons/pull/17478).

```
M215 - Get TOF Sensor Status
M222 - Get TOF Driver Register
M223 - Set TOF Driver Register
M224 - Enable TOF Driver
```

Closes: [EXEC-971](https://opentrons.atlassian.net/browse/EXEC-971) [EXEC-1166](https://opentrons.atlassian.net/browse/EXEC-1166) [EXEC-1170](https://opentrons.atlassian.net/browse/EXEC-1170) [EXEC-1171](https://opentrons.atlassian.net/browse/EXEC-1171) [EXEC-1172](https://opentrons.atlassian.net/browse/EXEC-1172)

## Test Plan and Hands on Testing

- [x] Make sure both sensors are initialized to the "Measure" application on cold boot.
- [x] Make sure you can get the TOF sensor status for both sensors
- [x] Make sure you can set and get the registers for each TOF sensor
- [x] Make sure you can enable/disable the TOF sensor

## Changelog

- Adds a "TOFSensor" task that is responsible for interfacing with the tmf8820 driver
- Adds a "TMF8820" driver interface class to load the measure application on cold boot.
- Added `scripts/hex_to_c.py` script that generates c array files from hex files.
- Added `scripts/requirements.txt` to keep track of script python dependencies, only `intelhex` for now.
- Added a cmake mechanism to install the `scripts/requirements.txt` file
- Added a cmake mechanism for the flex-stacker that uses the scripts/hex_to_c.py to generate the tmf8820_image.c file from the tmf8820_image.hex file.
- Set the I2C device address to custom addresses for both sensors (0x39 for X) and (0x40 for Z)
- add M215 - Get TOF Sensor Status Gcode and handling
- add M222 - Get TOF Driver Register Gcode and handling
- add M223 - Set TOF Driver Register Gcode and handling
- add M224 - Enable TOF Driver Gcode and handling


## Reviewer

Feedback on the cmake changes would be great.

## Risk assessment

Low, unreleased

[EXEC-971]: https://opentrons.atlassian.net/browse/EXEC-971?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1166]: https://opentrons.atlassian.net/browse/EXEC-1166?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1170]: https://opentrons.atlassian.net/browse/EXEC-1170?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1171]: https://opentrons.atlassian.net/browse/EXEC-1171?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1172]: https://opentrons.atlassian.net/browse/EXEC-1172?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
add new GetTOFSensorStatus (M215) gcode

vegano1 committed:
format

vegano1 committed:
add TOFSensorDriverTask scaffolding, GetTOFRegister, SetTOFRegister messages

vegano1 committed:
add TOFSensorTask scaffolding

vegano1 committed:
format

vegano1 committed:
save i2c hw

vegano1 committed:
add i2c_hardare

vegano1 committed:
add TOFDriverPolicy

vegano1 committed:
hook up SetTOFDriver gcode

vegano1 committed:
add transmit_receive i2c method

vegano1 committed:
use I2CComms instead of TOFDriverPolicy in TOFDriverTask

vegano1 committed:
add write protect functions for eeprom and tof sensors

vegano1 committed:
Merge branch 'edge' into tof-from-scratch

vegano1 committed:
Merge branch 'edge' into tof-from-scratch

vegano1 committed:
debug i2c line

vegano1 committed:
hallelujah! i2c comms work!

vegano1 committed:
move i2c setup to main

vegano1 committed:
save

vegano1 committed:
fixed the i2c line!
- switched to HAL_I2C_Mem_Read_IT and HAL_I2C_Mem_Write_IT, as i was having issues with Master Rear/Write IT
- added implementation for SetTOFRegister and GetTOFRegister, still need validation

vegano1 committed:
added EnableTOFSensor and implementation

vegano1 committed:
format

vegano1 committed:
rename tof_i2c_hardware and i2c files

vegano1 committed:
save

vegano1 committed:
start adding type to enum convertion for registers

vegano1 committed:
save

vegano1 committed:
Merge branch 'edge' into flex-stacker-tof-sensors-bringup

vegano1 committed:
fix bad merge

vegano1 committed:
tof

vegano1 committed:
remove annoying ^M carriedge return from from files.

vegano1 committed:
add startup tmf8821

vegano1 committed:
save

vegano1 committed:
Merge branch 'edge' into flex-stacker-tof-sensors-bringup

vegano1 committed:
feat(flex-stacker): enable tof sensors

sfoster1 commented on stm32-modules/flex-stacker/firmware/system/i2c_hardware.c:
```diff
@@ -1,5 +1,9 @@
 #include <stdint.h>
 
+#pragma GCC push_options
+#pragma GCC optimize("O0")
```
we sure about this one? What problem is it solving?

vegano1 replied:
nope, this was for debugging, will remove it.

sfoster1 commented on stm32-modules/flex-stacker/firmware/tof_sensor/tmf8820_image.c:
```diff
@@ -0,0 +1,227 @@
+/* http://srecord.sourceforge.net/ */
```
this will work, but I would love if we could do something like the following:
- Have the image in the repo as a binary, straight up
- do one of the following:
  - generate a source file like this from the binary image using a python script or using `xxd` or something, integrated as a cmake task, in a similar way to how we handle versions
  - pack it into the linked binary directly

sfoster1 commented on stm32-modules/flex-stacker/firmware/tof_sensor/tof_sensor_hardware.c:
```diff
@@ -0,0 +1,72 @@
+/**
+ * @brief enable the TOF X/Z write protect pin.
+ */
+void tof_write_protect_init(void) {
```
this is an enable, not a write protect, right?

sfoster1 commented on stm32-modules/flex-stacker/firmware/tof_sensor/tof_sensor_hardware.c:
```diff
@@ -0,0 +1,72 @@
+/**
+ * enable/disable writing to a tof sensor.
+ */
+void enable_tof_sensor_write(TOFSensorID sensor_id, bool enable) {
```
similarly, this puts the sensor into/takes it out of standby mode, right? It doesn't control writing.

sfoster1 commented on stm32-modules/flex-stacker/firmware/tof_sensor/tof_sensor_hardware.c:
```diff
@@ -0,0 +1,72 @@
+
+// Initialize the eeprom and tof enable pins
+void tof_hardware_init(void) {
+    eeprom_write_protect_init();
```
why are we binding the eeprom stuff into the tof stuff? they're not really connected, right? they're not even on the same bus?

vegano1 replied:
removed

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/host_comms_task.hpp:
```diff
@@ -67,33 +69,31 @@ class HostCommsTask {
           task_registry(aggregator),
           // These nolints are because if you don't have these inits, host
-          // builds complain NOLINTNEXTLINE(readability-redundant-member-init)
+          // builds conplain
```
```suggestion
          // builds complain
```

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+constexpr uint8_t BL_HEADER_LEN = 2;  // cmd + len
+constexpr uint8_t BL_FOOTER_LEN = 1;  // checksum
+constexpr uint8_t BL_BUFFER_LEN = BL_HEADER_LEN + BL_DATA_LEN + BL_FOOTER_LEN;
+uint8_t BUFFER[BL_BUFFER_LEN] = {0};
```
We don't want this in the header file; it's fine right now because I guess this header is only in one translation unit, but were we ever to include it in another translation unit we'd get linker duplicate-symbol errors because we're violating the one-definition rule. We can declare it here but we need to define it in a source file or inside a class def or something.

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+
+class TMF8821 {
+  public:
+    TMF8821(TOFSensorPolicy* policy) : _policy(policy) {}
```
why do we need the policy provided in both the ctor and the `initialize` function? 

also we should make any one-argument ctor `explicit`

vegano1 replied:
The policy in the ctor is initialized as a nullptr, then passed in the `initialize` function to set the pointer to the actual policy instance that we can reference within the class. Would it be better to pass in policy at compile time as a template parameter?

vegano1 replied:
removed from ctor since initializing as nullptr.

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+               int size = 1) -> std::optional<RegisterSerializedType> {
+        using RT = std::optional<RegisterSerializedType>;
+        auto dev_address = get_sensor_i2c_address(sensor_id);
+        auto [res, _] =
```
this syntax is so good

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+        auto dev_address = get_sensor_i2c_address(sensor_id);
+        auto [res, _] =
+            _policy->i2c_write(dev_address << 1, reg, (uint8_t*)data, size);
+        if (res != 0) return RT();
```
braces please

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+        using RT = std::optional<RegisterSerializedType>;
+        auto dev_address = get_sensor_i2c_address(sensor_id);
+        auto [res, data] = _policy->i2c_read(dev_address << 1, reg, size);
+        if (res != 0) return RT();
```
braces please

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+    // Gets the sensor i2c address
+    auto get_sensor_i2c_address(TOFSensorID sensor_id, bool default_addr = true)
+        -> uint16_t {
+        if (default_addr && !_custom_address) return TOF_DEFAULT_ADDRESS;
```
braces for all of these please

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+    auto reset_custom_address() -> void { _custom_address = false; }
+
+    // Gets the sensor i2c address
+    auto get_sensor_i2c_address(TOFSensorID sensor_id, bool default_addr = true)
```
I'm a little confused by the logic with having both the `default_addr` param and the built-in `_custom_address` field - why do we need both?

vegano1 replied:
Good call, `default_addr` is redundant.

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+
+        // Write the config page
+        auto len = prepare_cmd_frame(CMD_WRITE_CONFIG_PAGE, 0, 0);
+        if (!write(sensor_id, _registers.cmd_stat.address, (uint32_t*)BUFFER,
```
if this is what we're using `BUFFER` for then it definitely should be a class member. There's definitely going to be 2 of these class instances, which means that they could stomp on each other if they're sharing the same buffer. Putting it in the class will avoid that.

sfoster1 replied:
also the buffer size is defined in bytes and is `131` which definitely does not align with punning it to an array of uint32s, we'd get a guaranteed overflow if we ever wrote to that final word

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+    auto read_register(TOFSensorID sensor_id) -> std::optional<Reg> {
+        using RT = std::optional<Reg>;
+        auto ret = read(sensor_id, Reg::address, 1);
+        if (!ret.has_value()) return RT();
```
braces please

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        auto value = get_register_value(reg);
+        auto ret = write(sensor_id, Reg::address, &value, 1);
+        if (!ret.has_value()) return RT();
```
braces please

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp:
```diff
@@ -0,0 +1,457 @@
+                             ? BL_DATA_LEN
+                             : (tmf8820_image_length - i);
+            // send chunk
+            if (!bl_send_image_chunk(sensor_id, (uint8_t*)&tmf8820_image[i],
```
imo pointer arithmetic is a bit easier to follow than this but that's a light opinion

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tof_sensor_task.hpp:
```diff
@@ -0,0 +1,202 @@
+            _policy = policy;
+
+            // Disable both sensors before initializing
+            if (!_tof_sensor_x.ok) _policy->enable_tof_sensor(TOF_X, false);
```
braces please

sfoster1 reviewed (changes requested):
Some more minor things, but while the structure of this looks good we really shouldn't use a file-scope global as a shared buffer between two objects and I would love to see the tof firmware be included a little more nicely.

vegano1 committed:
- moved eepriom write pin hardware functions to system.c
- rename tof_write_protect to tof_enable_pin
- change BUFFER to be uint8_t instead of uint32_t.
- cleaning up

vegano1 committed:
format

vegano1 committed:
- added scripts/hex_to_c.py script that generates c array files from hex files.
- added scripts/requirements.txt to keep track of script python dependencies
- added a cmake mechanism to install the scripts/requirements.txt file
- added a cmake mechanism for the flex-stacker that uses the scripts/hex_to_c.py to generate the tmf8820_image.c file from the tmf8820_image.hex file.

vegano1 committed:
format

sfoster1 reviewed (approved):
Looks good to me!

ahiuchingau reviewed (approved):
The sensor hardware, tasks and policy structure looks good to me! Thank you

vegano1 committed:
Merge branch 'edge' into flex-stacker-tof-sensors-bringup
